File: src/git.clj
(ns git
  (:require
   const file path
   [clojure.string :as str]
   [command :as cmd]))

;; TODO: Figure out how to get git history of org-mode files as well. I don't want to lose it!
;; ASSUMES file is in relative path to wiki repo
(defn log [file source-dir]
  (let [file (str file)
        res (cmd/exec (str "git log --all --full-history --pretty=\"format:%h %H %ad\" --date default --date=format:'%Y-%m-%d' " (str file)) source-dir)]
    (if res
      (for [line (str/split res #"\R")]
        (let [[short-hash long-hash commit-date] (str/split line #" ")]
          {:short-hash short-hash
           :long-hash long-hash
           :commit-date commit-date
           :file-path file}))
      (throw (Throwable. (str "git log command failed on path " file))))))

;; ASSUMES file is in relative path to wiki repo
(defn last-log [file source-dir]
  (let [file (str file)
        res (cmd/exec (str "git log -1 --full-history --pretty=\"format:%h %H %ad %ct\" --date default --date=format:'%Y-%m-%d' " (str file)) source-dir)]
    (if res
      (let [line (first (str/split res #"\R"))
            [short-hash long-hash commit-date timestamp] (str/split line #" ")]
        {:short-hash short-hash
         :long-hash long-hash
         :commit-date commit-date
         :timestamp (Integer/parseInt timestamp)
         :file-path file})
      (throw (Throwable. (str "git log command failed on path " file))))))

(defn last-timestamp
  ([source-dir] (Integer/parseInt (cmd/exec (str  "git log -1  --pretty=format:%ct") source-dir)))
  ([source-dir path] (Integer/parseInt (cmd/exec (str  "git log -1  --pretty=format:%ct --follow -- " (str path)) source-dir))))

(defn history-link
  "Generate a link to git history"
  [long-hash file-path]
  (str const/source-url "/blob/" long-hash "/" file-path))

(defn checkout [source-dir branch]
  (cmd/exec (str "git checkout " branch) source-dir))

(defn add-all [source-dir]
  (cmd/exec (str "git add .") source-dir))

(defn commit [source-dir]
  (cmd/exec  "git -c commit.gpgsign=false commit -m \"robot commit\"" source-dir))

(defn push [source-dir]
  (cmd/exec  "git push" source-dir))

(defn remove-untracked [dir]
  (cmd/exec "git clean -fxd" dir))

(defn current-branch [dir]
  (cmd/exec "git branch --show-current" dir))

(defn status [dir]
  (println (cmd/exec "git status" dir)))

(defn stash [dir]
  (cmd/exec "git stash" dir))

(defn stash-pop [dir]
  (cmd/exec "git stash pop" dir))

File: src/git.js
const util = require('util');
const exec = util.promisify(require('child_process').exec);
const path = require('path');

async function log(file, sourceDir) {
  const filePath = path.join(sourceDir, file);
  const command = `git log --all --full-history --pretty="format:%h %H %ad" --date default --date=format:'%Y-%m-%d' ${filePath}`;
  try {
    const { stdout } = await exec(command);
    if (stdout) {
      return stdout.split('\n').map((line) => {
        const [shortHash, longHash, commitDate] = line.split(' ');
        return {
          shortHash,
          longHash,
          commitDate,
          file: file,
        };
      });
    } else {
      throw new Error(`git log command failed on path ${file}`);
    }
  } catch (error) {
    throw error;
  }
}

async function lastLog(file, sourceDir) {
  const filePath = path.join(sourceDir, file);
  const command = `git log -1 --full-history --pretty="format:%h %H %ad %ct" --date default --date=format:'%Y-%m-%d' ${filePath}`;
  try {
    const { stdout } = await exec(command);
    if (stdout) {
      const [line] = stdout.split('\n');
      const [shortHash, longHash, commitDate, timestamp] = line.split(' ');
      return {
        shortHash,
        longHash,
        commitDate,
        timestamp: parseInt(timestamp, 10),
        file: file,
      };
    } else {
      throw new Error(`git log command failed on path ${file}`);
    }
  } catch (error) {
    throw error;
  }
}

async function lastTimestamp(sourceDir, path) {
  const filePath = path ? path : '';
  const command = `git log -1 --pretty=format:%ct --follow -- ${filePath}`;
  try {
    const { stdout } = await exec(command, { cwd: sourceDir });
    return parseInt(stdout, 10);
  } catch (error) {
    throw error;
  }
}

function historyLink(longHash, filePath) {
  return `${const.sourceUrl}/blob/${longHash}/${filePath}`;
}

async function checkout(sourceDir, branch) {
  await exec(`git checkout ${branch}`, { cwd: sourceDir });
}

async function addAll(sourceDir) {
  await exec('git add .', { cwd: sourceDir });
}

async function commit(sourceDir) {
  await exec('git -c commit.gpgsign=false commit -m "robot commit"', { cwd: sourceDir });
}

async function push(sourceDir) {
  await exec('git push', { cwd: sourceDir });
}

async function removeUntracked(dir) {
  await exec('git clean -fxd', { cwd: dir });
}

async function currentBranch(dir) {
  const { stdout } = await exec('git branch --show-current', { cwd: dir });
  return stdout.trim();
}

async function status(dir) {
  const { stdout } = await exec('git status', { cwd: dir });
  console.log(stdout);
}

async function stash(dir) {
  await exec('git stash', { cwd: dir });
}

async function stashPop(dir) {
  await exec('git stash pop', { cwd: dir });
}

module.exports = {
  log,
  lastLog,
  lastTimestamp,
  historyLink,
  checkout,
  addAll,
  commit,
  push,
  removeUntracked,
  currentBranch,
  status,
  stash,
  stashPop,
};
